### ISSUE TITLE
Exception during building of documentation of pep8 1.5.6

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1467]
Exception occurs during building of documentation of pep8 1.5.6
(https://pypi.python.org/pypi/pep8 ; https://pypi.python.org/packages/source/p/pep8/pep8-1.5.6.tar.gz).

This exception occurs when using Sphinx with Python 3, not Python 2.

This exception occurs with both Sphinx 1.2.2 and trunk.

Result with current trunk (with -v option, so full traceback is already shown):

```
Running Sphinx v1.3a0
loading pickled environment... failed: [Errno 2] No such file or directory: '/tmp/pep8-1.5.6/docs/build/.doctrees/environment.pickle'
No builder selected, using default: html
building [html]: targets for 5 source files that are out of date
updating environment: 5 added, 0 changed, 0 removed
reading sources... [ 20&#37;] advanced
reading sources... [ 40&#37;] api

Traceback (most recent call last):
  File "/tmp/sphinx/build/lib/sphinx/cmdline.py", line 253, in main
    app.build(force_all, filenames)
  File "/tmp/sphinx/build/lib/sphinx/application.py", line 214, in build
    self.builder.build_update()
  File "/tmp/sphinx/build/lib/sphinx/builders/__init__.py", line 214, in build_update
    'out of date' &#37; len(to_build))
  File "/tmp/sphinx/build/lib/sphinx/builders/__init__.py", line 234, in build
    purple, length):
  File "/tmp/sphinx/build/lib/sphinx/builders/__init__.py", line 134, in status_iterator
    for item in iterable:
  File "/tmp/sphinx/build/lib/sphinx/environment.py", line 474, in update_generator
    self.read_doc(docname, app=app)
  File "/tmp/sphinx/build/lib/sphinx/environment.py", line 621, in read_doc
    pub.publish()
  File "/usr/lib64/python3.4/site-packages/docutils/core.py", line 217, in publish
    self.settings)
  File "/usr/lib64/python3.4/site-packages/docutils/readers/__init__.py", line 72, in read
    self.parse()
  File "/usr/lib64/python3.4/site-packages/docutils/readers/__init__.py", line 78, in parse
    self.parser.parse(self.input, document)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/__init__.py", line 172, in parse
    self.statemachine.run(inputlines, document, inliner=self.inliner)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 170, in run
    input_source=document['source'])
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2961, in text
    self.section(title.lstrip(), source, style, lineno + 1, messages)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 327, in section
    self.new_subsection(title, lineno, messages)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 395, in new_subsection
    node=section_node, match_titles=True)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 282, in nested_parse
    node=node, match_titles=match_titles)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 195, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2726, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 327, in section
    self.new_subsection(title, lineno, messages)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 395, in new_subsection
    node=section_node, match_titles=True)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 282, in nested_parse
    node=node, match_titles=match_titles)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 195, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2301, in explicit_markup
    self.explicit_list(blank_finish)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2331, in explicit_list
    match_titles=self.state_machine.match_titles)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 319, in nested_list_parse
    node=node, match_titles=match_titles)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 195, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2604, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2311, in explicit_construct
    return method(self, expmatch)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2054, in directive
    directive_class, match, type_name, option_presets)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2103, in run_directive
    result = directive_instance.run()
  File "/tmp/sphinx/build/lib/sphinx/ext/autodoc.py", line 1458, in run
    self.state.nested_parse(self.result, 0, node)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 282, in nested_parse
    node=node, match_titles=match_titles)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 195, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2299, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2311, in explicit_construct
    return method(self, expmatch)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2054, in directive
    directive_class, match, type_name, option_presets)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2103, in run_directive
    result = directive_instance.run()
  File "/tmp/sphinx/build/lib/sphinx/domains/__init__.py", line 190, in run
    return BaseDirective.run(self)
  File "/tmp/sphinx/build/lib/sphinx/directives/__init__.py", line 155, in run
    self.state.nested_parse(self.content, self.content_offset, contentnode)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 282, in nested_parse
    node=node, match_titles=match_titles)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 195, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2301, in explicit_markup
    self.explicit_list(blank_finish)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2331, in explicit_list
    match_titles=self.state_machine.match_titles)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 319, in nested_list_parse
    node=node, match_titles=match_titles)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 195, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/lib64/python3.4/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2604, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2311, in explicit_construct
    return method(self, expmatch)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2054, in directive
    directive_class, match, type_name, option_presets)
  File "/usr/lib64/python3.4/site-packages/docutils/parsers/rst/states.py", line 2103, in run_directive
    result = directive_instance.run()
  File "/tmp/sphinx/build/lib/sphinx/ext/autodoc.py", line 1433, in run
    documenter.generate(more_content=self.content)
  File "/tmp/sphinx/build/lib/sphinx/ext/autodoc.py", line 750, in generate
    if not self.import_object():
  File "/tmp/sphinx/build/lib/sphinx/ext/autodoc.py", line 1218, in import_object
    obj_from_parent = self.parent.__dict__.get(self.object_name)
AttributeError: 'NoneType' object has no attribute '__dict__'

Exception occurred:
  File "/tmp/sphinx/build/lib/sphinx/ext/autodoc.py", line 1218, in import_object
    obj_from_parent = self.parent.__dict__.get(self.object_name)
AttributeError: 'NoneType' object has no attribute '__dict__'
The full traceback has been saved in /tmp/sphinx-err-yyzbvh45.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://bitbucket.org/birkenfeld/sphinx/issues/>. Thanks!
```

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1467
- Originally reported by: Arfrever Frehtes Taifersar Arahesis
- Originally created at: 2014-05-12T19:26:17.222

**Comment by shimizukawa** [comment]
_From Arfrever Frehtes Taifersar Arahesis on 2014-05-31 10:51:22+00:00_

Problem was fixed at pep8 side in https://github.com/jcrocholl/pep8/commit/b14a97e5753b9c63794b76a8ef66ef9a03186b79 (released in pep8 1.5.7), but Sphinx still needs to be fixed to print informative message instead of traceback.

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-06-08 00:17:59+00:00_

`pep8.Checker.readline_check_physical` method has been removed at 
https://github.com/jcrocholl/pep8/commit/fed43c5b560c90995e2ea3f742f2bf6b6e497657 but `api.rst` had referred it. it's a cause.

However, as you mentioned, sphinx on python2 and on python3 has a difference behavior. Sphinx on py2 will emit WARNING in such case but sphinx on py3 will raise exception. I'll fix it.

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-06-08 16:52:15+00:00_
- Fix exception on Python3 if nonexistent method is specified by automethod. Closes #1467

→ <<cset 1d742bdcf5181c0b8b4a3be310d28179d3d14819>>
